/*!********************************************************************************
 * \brief     follow_path implementation
 * \authors   Pablo Santamaria
 * \copyright Copyright (c) 2021 Universidad Politecnica de Madrid
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holder nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *******************************************************************************/

#ifndef FOLLOW_PATH_H
#define FOLLOW_PATH_H

#include <yaml-cpp/yaml.h>
// ROS
#include "std_msgs/Float32MultiArray.h"
#include <ros/ros.h>
#include <std_msgs/Bool.h>

// Aerostack msgs
#include <behavior_execution_manager_msgs/BehaviorActivationFinished.h>
#include <aerostack_msgs/FlightActionCommand.h>
#include "aerostack_msgs/FlightState.h"
#include <nav_msgs/Path.h>
#include "trajectory_msgs/JointTrajectoryPoint.h"
// Aerostack libraries
#include <BehaviorExecutionManager.h>
#include "ros_utils_lib/ros_utils.hpp"


class BehaviorFollowPathWithDF : public BehaviorExecutionManager
{
  // Constructor
public:
  BehaviorFollowPathWithDF();
  ~BehaviorFollowPathWithDF();

private:
  // Congfig variables
  std::string command_high_level_str;
  std::string status_str;
  std::string path_blocked_topic_str;
  std::string motion_reference_traj_topic_;
  
  ros::NodeHandle node_handle;
  std::string nspace; 
  // Subscriber
  ros::Subscriber status_sub;
  ros::Subscriber path_blocked_sub;
  ros::Subscriber traj_sub_;
  //Publishers
  ros::Publisher command_high_level_pub;
  

  // Messages
  aerostack_msgs::FlightState status_msg;
  aerostack_msgs::FlightActionCommand high_level_command;

  bool path_blocked;

private:
  // BehaviorExecutionManager
  void onConfigure();
  void onActivate();
  void onDeactivate();
  void onExecute();
  bool checkSituation();
  void checkGoal();
  void checkProgress();
  void checkProcesses();

public: 
// Callbacks
void statusCallBack(const aerostack_msgs::FlightState &msg);
void pathBlockedCallBack(const std_msgs::Bool &msg);
void CallbackTrajectoryTopic(const trajectory_msgs::JointTrajectoryPoint& traj);

};

#endif
